File: bench/arrow.mjs
// What a result costs as Arrow, against the two other ways out.
//
// Three calls over the same statement. `query` builds an object a row
// and a value a cell. `columnar` moves one buffer a column and leaves
// the reader to put a type around them. `arrow` writes those same
// buffers into an IPC stream, which costs a copy of the values and a
// header a batch, and buys a result every Arrow implementation reads.
//
// So the pair worth reading is `arrow` against `columnar`: the
// difference between them is what the framing costs, and it is the
// number that says whether a caller should take the bytes or take the
// buffers. The line after each is the reader's side, since bytes nobody
// decodes are not a result: `tableFromIPC` against the eleven lines the
// README prints for wrapping the buffers.
//
// Run it against a release build, for the reason bench/query.mjs gives.
//
//   npm run build && npm run bench:arrow

import { mkdtemp, rm } from 'node:fs/promises'
import { tmpdir } from 'node:os'
import { join } from 'node:path'

import { tableFromIPC } from 'apache-arrow'
import { connect } from 'zudb'

const ROWS = Number(process.env.ZU_BENCH_ROWS ?? 1_000_000)
const REPEATS = Number(process.env.ZU_BENCH_REPEATS ?? 5)

const dir = await mkdtemp(join(tmpdir(), 'zu-bench-arrow-'))
const conn = await connect(join(dir, 'bench.zu1'))

await conn.exec("INSERT (p:person {uid: 1, score: 1.5, name: 'n1'})")
{
  const rows = await conn.appender('person')
  for (let ix = 2; ix <= ROWS; ix++) rows.appendRow([BigInt(ix), ix / 3, `n${ix}`])
  await rows.close()
}

// The fastest of `REPEATS` runs, in milliseconds, after one warmup, for
// the reason bench/query.mjs gives: everything that makes a run slower
// than the work itself is something that happened to it rather than
// something about it.
async function time(run) {
  await run()
  let best = Infinity
  for (let round = 0; round < REPEATS; round++) {
    const started = performance.now()
    await run()
    best = Math.min(best, performance.now() - started)
  }
  return best
}

function report(name, ms) {
  const each = (ms * 1e6) / ROWS
  console.log(
    `${name.padEnd(30)} ${ms.toFixed(1).padStart(8)} ms  ${Math.round(each).toString().padStart(6)} ns/row`,
  )
}

const ONE = 'MATCH (p:person) RETURN p.uid AS uid'
const THREE = 'MATCH (p:person) RETURN p.uid AS uid, p.score AS score, p.name AS name'

const cases = [
  { name: 'one integer column, arrow', run: () => conn.arrow(ONE) },
  { name: 'one integer column, columnar', run: () => conn.columnar(ONE) },
  { name: 'one integer column, rows', run: () => conn.query(ONE) },
  { name: 'three columns, arrow', run: () => conn.arrow(THREE) },
  { name: 'three columns, columnar', run: () => conn.columnar(THREE) },
  { name: 'three columns, rows', run: () => conn.query(THREE) },
]

console.log(`reading ${ROWS} rows, fastest of ${REPEATS}`)
for (const { name, run } of cases) report(name, await time(run))

// What a caller does next, which is the half the calls above do not
// include. The bytes are read once here and decoded every round, so what
// is timed is the decode and not the statement.
const bytes = (await conn.arrow(THREE)).ipc

console.log('')
console.log('turning what came back into a table')
report('tableFromIPC over the bytes', await time(async () => tableFromIPC(bytes)))

// A batch is a slice of arrays that are already built, so the size is
// about what the reader holds at once rather than about the write. This
// says by how much, which is the answer to whether it is worth tuning.
console.log('')
console.log('the batch size the stream is cut into')
for (const batchRows of [4_096, 65_536, 1_000_000]) {
  report(`batchRows ${batchRows}`, await time(() => conn.arrow(THREE, null, { batchRows })))
}

await conn.close()
await rm(dir, { recursive: true, force: true })

File: src/arrow.rs
//! A result as Arrow, in the bytes Arrow ships between processes.
//!
//! ```js
//! import { tableFromIPC } from 'apache-arrow'
//! const read = await conn.arrow('MATCH (p:person) RETURN p.name AS name, p.age AS age')
//! const table = tableFromIPC(read.ipc)
//! ```
//!
//! [`columnar`] hands over the buffers themselves and leaves the reader
//! to put a type around them, which is the fastest way out and the one
//! that costs a caller ten lines of Arrow before they have a table. This
//! is the other way: the same buffers, with the schema written beside
//! them, in the format every Arrow implementation already reads. What
//! comes back is bytes, so `apache-arrow` reads it, DuckDB-Wasm reads
//! it, a `fetch` response body carries it, and a worker gets it as a
//! transferable rather than as a structured clone.
//!
//! The translation is not written here. `zu-arrow` in the engine tree is
//! the one answer about what a zu column becomes in Arrow, shared with
//! the Python client, because a second copy of it would be a second set
//! of rules about what a year-month duration is. This module is the
//! runtime's half: read the option, run the statement, hand the bytes to
//! V8 without copying them again.
//!
//! Why bytes and not the C Data Interface, which is what the Python
//! client takes and is a pointer rather than a serialization: nothing in
//! a JavaScript runtime can dereference a pointer. An addon can, and
//! this one does on the way in, but the value that reaches JavaScript
//! has to be something V8 holds, and the only thing V8 holds that Arrow
//! also speaks is a buffer of IPC bytes. The framing is the cost: a
//! schema message, then a header a batch, which is kilobytes against a
//! result of any size and the price of a format with readers.
//!
//! [`columnar`]: crate::columns
use napi::bindgen_prelude::*;
use napi::{Env, ScopedTask};
use zudb::DiagnosticRecord;

use crate::conn::{Failure, QueryTask, notices};

/// One statement, read as Arrow.
pub struct ArrowTask {
    pub(crate) task: QueryTask,
    /// How many rows go in a record batch, or why the caller's answer to
    /// that could not be read.
    pub(crate) batch: std::result::Result<usize, String>,
}

/// A whole result, as the bytes and what came with them.
pub struct Read {
    ipc: Vec<u8>,
    rows: usize,
    gqlstatus: &'static str,
    notices: Vec<DiagnosticRecord>,
}

impl ArrowTask {
    fn run(&mut self) -> std::result::Result<Read, Failure> {
        // Before the statement, because a batch size nobody could read
        // is a call that was wrong when it was written and not an answer
        // worth running a scan for.
        let batch = self
            .batch
            .as_ref()
            .map_err(|message| Failure::Usage(message.clone()))?;
        let batch = *batch;
        let (result, shape) = self.task.run()?;
        // The names are the ones the statement's own catalog gave, so a
        // node column names its table rather than its id.
        let ipc = zu_arrow::ipc(&result, shape.names(), batch)
            .map_err(|err| Failure::Usage(err.to_string()))?;
        Ok(Read {
            ipc,
            // Answered off the columns the sink filled, which is a read
            // of a length and not a pass that builds rows nobody wants.
            rows: result.rows.len(),
            gqlstatus: result.status().code(),
            notices: result.notices,
        })
    }
}

impl<'task> ScopedTask<'task> for ArrowTask {
    type Output = std::result::Result<Read, Failure>;
    type JsValue = Object<'task>;

    fn compute(&mut self) -> Result<Self::Output> {
        Ok(self.run())
    }

    fn resolve(&mut self, env: &'task Env, output: Self::Output) -> Result<Self::JsValue> {
        let read = output.map_err(|failure| self.task.failed(env, failure))?;
        let mut object = Object::new(env)?;
        // The `Vec` is moved and not read: the pointer V8 is handed is
        // the pointer the writer filled, and the allocation is freed when
        // the typed array is collected.
        object.set("ipc", Uint8Array::new(read.ipc))?;
        object.set("rows", read.rows as f64)?;
        object.set("gqlstatus", read.gqlstatus)?;
        object.set("notices", notices(env, &read.notices)?)?;
        Ok(object)
    }

    fn finally(mut self, env: Env) -> Result<()> {
        self.task.release(&env)
    }
}
